### PR TITLE
exclude exports

### PR DESCRIPTION
**Pull request by labhvam5**

### CODE REVIEW

File: apps/fyle/helpers.py [hunk 1]
```diff
@@ -6,8 +6,9 @@
 
 from django.utils.module_loading import import_string
 from django.conf import settings
+from django.db.models import Q
 
-from apps.fyle.models import ExpenseGroupSettings
+from apps.fyle.models import ExpenseGroupSettings,ExpenseFilter
```
Comment (NileshPant1999) [inline apps/fyle/helpers.py:11]: space after `,` will be good to maintain consistency

File: apps/fyle/helpers.py [hunk 2]
```diff
@@ -185,4 +186,32 @@ def sync_dimensions(fyle_credentials: FyleCredential, workspace_id: int) -> None
 
     platform.import_fyle_dimensions(import_taxes=True)
    
-    
+
+def construct_expense_filter(expense_filter:ExpenseFilter):
```
Comment (Shwetabhk) [inline apps/fyle/helpers.py:190]:
```suggestion
def construct_expense_filter(expense_filter: ExpenseFilter):
```

File: apps/fyle/helpers.py (hunk 1 above)
Comment (Shwetabhk) [inline apps/fyle/helpers.py:11]:
```suggestion
from apps.fyle.models import ExpenseGroupSettings, ExpenseFilter
```

File: apps/fyle/helpers.py [hunk 3]
```diff
@@ -185,4 +186,32 @@ def sync_dimensions(fyle_credentials: FyleCredential, workspace_id: int) -> None
...
+    constructed_expense_filter = {}
```
Comment (Shwetabhk) [inline apps/fyle/helpers.py:191]: Please add some comments about each if condition in the code

File: apps/fyle/helpers.py [hunk 4]
```diff
@@ -185,4 +186,32 @@ def sync_dimensions(fyle_credentials: FyleCredential, workspace_id: int) -> None
...
+    if expense_filter.is_custom and expense_filter.operator != 'isnull':
+        filter1 = {
```
Comment (Shwetabhk) [inline apps/fyle/helpers.py:193]:
```suggestion
        # This block is executed when for custom fields null check
        # Eg. (Add a constructed condition here @labhvam5 )
        filter1 = {
```

File: apps/fyle/helpers.py [hunk 5]
```diff
@@ -185,4 +186,32 @@ def sync_dimensions(fyle_credentials: FyleCredential, workspace_id: int) -> None
...
+            'custom_properties__{0}__exact'.format(expense_filter.condition): None
+        }
+        if expense_filter_value == True:
+            constructed_expense_filter = Q(**filter1) | Q(**filter2)
+        else:
+            constructed_expense_filter = ~Q(**filter2)
+
+    else:
+        filter1 = {
+            '{0}__{1}'.format(expense_filter.condition, expense_filter.operator):expense_filter.values if len(expense_filter.values) > 1 else expense_filter.values[0]
```
Comment (Shwetabhk) [inline apps/fyle/helpers.py:213]: Line too long, need to shorten it
```suggestion
            '{0}__{1}'.format(
                expense_filter.condition, 
                expense_filter.operator
            ):expense_filter.values if len(expense_filter.values) > 1 else expense_filter.values[0]
```

File: apps/fyle/models.py [hunk 6]
```diff
@@ -436,3 +462,23 @@ def get_last_synced_at(workspace_id: int):
...
+class ExpenseFilter(models.Model):
+    """
+    Reimbursements
+    """
+    id = models.AutoField(primary_key=True)
+    condition = models.CharField(max_length=255, help_text='Condition for the filter')
+    operator = models.CharField(max_length=255,choices=EXPENSE_FILTER_OPERATOR, help_text='Operator for the filter')
+    values = ArrayField(base_field=models.CharField(max_length=255),null=True ,help_text='Values for the operator')
+    rank = models.IntegerField(choices=EXPENSE_FILTER_RANK, help_text='Rank for the filter')
+    join_by = models.CharField(max_length=255,null=True,choices=EXPENSE_FILTER_JOIN_BY, help_text='Used to join the filter (AND/OR)')
```
Comment (Shwetabhk) [inline apps/fyle/models.py:476]: break into two lines

File: apps/fyle/models.py [hunk 7]
```diff
@@ -436,3 +462,23 @@ def get_last_synced_at(workspace_id: int):
...
+
+
+class ExpenseFilter(models.Model):
+    """
+    Reimbursements
+    """
+    id = models.AutoField(primary_key=True)
+    condition = models.CharField(max_length=255, help_text='Condition for the filter')
+    operator = models.CharField(max_length=255,choices=EXPENSE_FILTER_OPERATOR, help_text='Operator for the filter')
+    values = ArrayField(base_field=models.CharField(max_length=255),null=True ,help_text='Values for the operator')
```
Comment (Shwetabhk) [inline apps/fyle/models.py:474]:
```suggestion
    values = ArrayField(base_field=models.CharField(max_length=255), null=True , help_text='Values for the operator')
```

File: apps/fyle/models.py [hunk 8]
```diff
@@ -436,3 +462,23 @@ def get_last_synced_at(workspace_id: int):
...
         ).order_by('-updated_at').first()
+
+
+class ExpenseFilter(models.Model):
+    """
+    Reimbursements
+    """
+    id = models.AutoField(primary_key=True)
+    condition = models.CharField(max_length=255, help_text='Condition for the filter')
+    operator = models.CharField(max_length=255,choices=EXPENSE_FILTER_OPERATOR, help_text='Operator for the filter')
```
Comment (Shwetabhk) [inline apps/fyle/models.py:473]:
```suggestion
    operator = models.CharField(max_length=255, choices=EXPENSE_FILTER_OPERATOR, help_text='Operator for the filter')
```

File: apps/fyle/models.py (hunk 6 above)
Comment (Shwetabhk) [inline apps/fyle/models.py:476]:
```suggestion
    join_by = models.CharField(
        max_length=255, null=True, 
        choices=EXPENSE_FILTER_JOIN_BY, 
        help_text='Used to join the filter (AND/OR)'
     )
```

File: apps/fyle/views.py [hunk 9]
```diff
@@ -254,3 +254,29 @@ def post(self, request, *args, **kwargs):
...
             )
+
+
+class ExpenseFilterView(generics.ListCreateAPIView):
+    """
+    Expense Filter view
+    """
+    serializer_class = ExpenseFilterSerializer
+
+    def get(self, request, *args, **kwargs):
```
Comment (Shwetabhk) [inline apps/fyle/views.py:265]: This isn't correct, if we define the `get` function then the whole point of using generic view goes away
```suggestion
    class ExpenseFilterView(generics.ListCreateAPIView):
    """
    Expense Filter view
    """
    serializer_class = ExpenseFilterSerializer
    
    def get_queryset(self):
        queryset = ExpenseFilter.objects.filter(workspace_id=self.kwargs['workspace_id'])
        return queryset
```

That's all, we don't need the whole function you've written

**Review by Shwetabhk — DISMISSED** [review]

File: apps/fyle/views.py [hunk 10]
```diff
@@ -254,3 +256,89 @@ def post(self, request, *args, **kwargs):
...
+class CustomFieldView(generics.RetrieveAPIView):
+    """
+    Custom Field view
+    """
+    def get(self, request, *args, **kwargs):
+        """
+        Get Custom Fields
+        """
+        try:
+            client_id = settings.FYLE_CLIENT_ID
```
Comment (Shwetabhk) [inline apps/fyle/views.py:306]: move all of this code to helpers.py

File: apps/fyle/views.py [hunk 11]
```diff
@@ -254,3 +256,89 @@ def post(self, request, *args, **kwargs):
...
+            client_secret = settings.FYLE_CLIENT_SECRET
+            token_url = settings.FYLE_TOKEN_URI
+            fyle_credentials = FyleCredential.objects.get(workspace_id=self.kwargs['workspace_id'])
+
+            fyle = Platform(
```
Comment (Shwetabhk) [inline apps/fyle/views.py:311]: All of this should be a part of PlatformConnector

File: apps/fyle/views.py [hunk 12]
```diff
@@ -254,3 +256,89 @@ def post(self, request, *args, **kwargs):
...
+
+        filters = {
+            'org_id': org_id,
+            'is_skipped': True
+        }
+
+        if start_date and end_date:
+            filters['created_at__range'] = [start_date, end_date]
+
+        queryset = Expense.objects.filter(**filters).values('created_at', 'expense_number', 'employee_email', 'employee_name', 'fund_source').order_by('created_at')
```
Comment (Shwetabhk) [inline apps/fyle/views.py:292]: The order by should be in descending order

Comment (Shwetabhk) [inline apps/fyle/views.py:292]: Since you've already specified fields in serializer I am not sure if you need to specify values here too, please verify once

Comment (labhvam5) [reply apps/fyle/views.py:292]: we don't need to specify fields in the queryset.

**Review by Shwetabhk — APPROVED** [review]

File: apps/fyle/tasks.py [hunk 13]
```diff
@@ -104,8 +105,16 @@ def create_expense_groups(workspace_id: int, configuration: Configuration, fund_
...
             expense_objects = Expense.create_expense_objects(expenses)
 
+            expenses_object_ids = [expense_object.id for expense_object in expense_objects]
+
+            expense_filters = ExpenseFilter.objects.filter(workspace_id=workspace_id)
+            if expense_filters:
+                final_query = construct_expense_filter_query(expense_filters)
+                Expense.objects.filter(final_query, id__in=expenses_object_ids, expensegroup__isnull=True, org_id=workspace.fyle_org_id).update(is_skipped=True)
+
+            filtered_expenses = Expense.objects.filter(is_skipped=False, id__in=expenses_object_ids, expensegroup__isnull=True, org_id=workspace.fyle_org_id)
```
Comment (ashwin1111) [inline apps/fyle/tasks.py:115]: can we just assign the value of `expense_objects` to `filtered_expenses` in the else block of 111th line? this query seems unnecessary, please also think of all edge cases

Comment (ashwin1111) [reply apps/fyle/tasks.py:115]: if you do this, in the if block, assign value to filtered expense

File: apps/fyle/views.py [hunk 14]
```diff
@@ -254,3 +261,81 @@ def post(self, request, *args, **kwargs):
...
+
+class ExpenseView(generics.ListAPIView):
+    """
+    Expense view
+    """
+
+    serializer_class = ExpenseSerializer
+
+    def get_queryset(self):
+        start_date = self.request.query_params.get('start_date', None)
```
Comment (ashwin1111) [inline apps/fyle/views.py:285]: why is this needed?

Comment (ashwin1111) [reply apps/fyle/views.py:285]: okay, for new design this might be required, okay to keep it

File: apps/fyle/views.py [hunk 15]
```diff
@@ -254,3 +261,81 @@ def post(self, request, *args, **kwargs):
...
+            'org_id': org_id,
+            'is_skipped': True
+        }
+
+        if start_date and end_date:
+            filters['updated_at__range'] = [start_date, end_date]
+
+        queryset = Expense.objects.filter(**filters).order_by('-updated_at')
+
+        return queryset
```
Comment (ashwin1111) [inline apps/fyle/views.py:299]: we'll need to support pagination for this API

File: sql/functions/delete-workspace.sql [hunk 16]
```diff
@@ -101,6 +101,12 @@ BEGIN
     GET DIAGNOSTICS rcount = ROW_COUNT;
     RAISE NOTICE 'Deleted % expenses', rcount;
 
+    DELETE
+    FROM expenses 
+    WHERE is_skipped=true and org_id = (SELECT fyle_org_id FROM workspaces WHERE id=_workspace_id);
```
Comment (ashwin1111) [inline sql/functions/delete-workspace.sql:106]: why? delete workspace is meant to delete whole org data

File: apps/fyle/migrations/0021_auto_20221222_1447.py [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+
+    operations = [
+        migrations.AddField(
+            model_name='expense',
+            name='is_skipped',
+            field=models.BooleanField(default=False, help_text='Expense is skipped or not', null=True),
+        ),
+        migrations.AddField(
+            model_name='expense',
+            name='report_title',
```
[the remark this answers is not in the document]
Comment (labhvam5) [reply apps/fyle/migrations/0021_auto_20221222_1447.py:23]: We don't need this.

File: apps/fyle/migrations/0022_expense_employee_name.py [hunk 18]
```diff
@@ -0,0 +1,18 @@
...
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('fyle', '0021_auto_20221222_1447'),
+    ]
+
+    operations = [
+        migrations.AddField(
+            model_name='expense',
+            name='employee_name',
```
[the remark this answers is not in the document]
Comment (labhvam5) [reply apps/fyle/migrations/0022_expense_employee_name.py:15]: We don't need this.

File: apps/fyle/models.py [hunk 19]
```diff
@@ -436,3 +464,36 @@ def get_last_synced_at(workspace_id: int):
...
+    values = ArrayField(base_field=models.CharField(max_length=255), null=True, help_text='Values for the operator')
+    rank = models.IntegerField(choices=EXPENSE_FILTER_RANK, help_text='Rank for the filter')
+    join_by = models.CharField(
+        max_length=255,
+        null=True,
+        choices=EXPENSE_FILTER_JOIN_BY,
+        help_text='Used to join the filter (AND/OR)'
+    )
+    is_custom = models.BooleanField(default=False, help_text='Custom Field or not')
+    custom_field_type = models.CharField(
```
[the remark this answers is not in the document]
Comment (labhvam5) [reply apps/fyle/models.py:485]: Yes we can.

File: sql/functions/delete-workspace.sql (hunk 16 above)
Comment (labhvam5) [reply sql/functions/delete-workspace.sql:106]: The `skipped` expenses don't get any `expense_groups` assigned to them . So we can identify the `skipped` expenses based on the `org_id` .

File: apps/fyle/helpers.py [hunk 20]
```diff
@@ -184,3 +186,64 @@ def sync_dimensions(fyle_credentials: FyleCredential, workspace_id: int) -> None
...
+    if expense_filter.is_custom and expense_filter.operator != 'isnull':
+        #This block is for custom-field with not null check 
+        if expense_filter.custom_field_type == 'NUMBER':
+            expense_filter.values = [int(expense_filter_value) for expense_filter_value in expense_filter.values]
+
+        filter1 = {
+            'custom_properties__{0}__{1}'.format(
+                expense_filter.condition,
+                expense_filter.operator
+            ): expense_filter.values if len(expense_filter.values) > 1 or expense_filter.operator == 'in' else expense_filter.values[0]
```
[the remark this answers is not in the document]
Comment (labhvam5) [reply apps/fyle/helpers.py:216]: No. 
- if `values = ['hello', 'hi', 'namaste' ]` then it should be should be ['hello', 'hi', 'namaste']. 
- Else if `value=['True'] then it should be `True` . 
- With the only exception of the `operator == in` in that case `['hello', 'hi', 'namaste']` will be `['hello', 'hi', 'namaste']` and `['hi`]` will be  `['hi']`

File: apps/fyle/helpers.py [hunk 21]
```diff
@@ -184,3 +186,64 @@ def sync_dimensions(fyle_credentials: FyleCredential, workspace_id: int) -> None
...
+            #if isnull=False
+            constructed_expense_filter = ~Q(**filter2)
+
+    else:
+        #This block is for all the non-custom-fields
+        filter1 = {
+            '{0}__{1}'.format(
+                expense_filter.condition,
+                expense_filter.operator
+            ):expense_filter.values if len(expense_filter.values) > 1 or expense_filter.operator == 'in' else expense_filter.values[0]
```
[the remark this answers is not in the document]
Comment (labhvam5) [reply apps/fyle/helpers.py:245]: No.

- if values = ['hello', 'hi', 'namaste' ] then it should be should be ['hello', 'hi', 'namaste'].

- Else if value=['True'] then it should be True .

- With the only exception of the operator == in in that case ['hello', 'hi', 'namaste'] will be ['hello', 'hi', 'namaste'] and ['hi'] will be ['hi']

File: apps/fyle/helpers.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (ashwin1111) [reply apps/fyle/helpers.py:216]: why do you need a len() comparision when you can compare with operator == 'in' ?

File: apps/fyle/helpers.py [hunk 22]
```diff
@@ -184,3 +186,64 @@ def sync_dimensions(fyle_credentials: FyleCredential, workspace_id: int) -> None
...
+    elif expense_filter.is_custom and expense_filter.operator == 'isnull':
+        #This block is for custom-field is null check
+        expense_filter_value: bool = True if expense_filter.values[0].lower() == 'true' else False
+        filter1 = {
+            'custom_properties__{0}__{1}'.format(
+                expense_filter.condition,
+                expense_filter.operator
+            ): expense_filter_value
+        }
+        filter2 = {
```
[the remark this answers is not in the document]
Comment (labhvam5) [reply apps/fyle/helpers.py:229]: yes. For `is_null = True` we need to account for `""` (empty string) and `null` values. 

- `isnull = true` will take care of `""` (empty string).
- `_exact = None` will take care of `null` values.

File: apps/fyle/tasks.py (hunk 13 above)
Comment (labhvam5) [reply apps/fyle/tasks.py:115]: Something like this ? 
```suggestion
            expense_objects = Expense.create_expense_objects(expenses)
            expense_filters = ExpenseFilter.objects.filter(workspace_id=workspace_id)

            if expense_filters:
                expenses_object_ids = [expense_object.id for expense_object in expense_objects]
                final_query = construct_expense_filter_query(expense_filters)
                Expense.objects.filter(final_query, id__in=expenses_object_ids, expensegroup__isnull=True, org_id=workspace.fyle_org_id).update(is_skipped=True)
                filtered_expenses = Expense.objects.filter(is_skipped=False, id__in=expenses_object_ids, expensegroup__isnull=True, org_id=workspace.fyle_org_id)
                
                ExpenseGroup.create_expense_groups_by_report_id_fund_source(
                    filtered_expenses, configuration, workspace_id
                )
            else:
                ExpenseGroup.create_expense_groups_by_report_id_fund_source(
                    expense_objects, configuration, workspace_id
                )
```

Comment (ashwin1111) [reply apps/fyle/tasks.py:115]: yes. `ExpenseGroup.create_expense_groups_by_report_id_fund_source()` can keep this out of block. In the else block you can just assign filtered_expenses = expense_objects

**Review by ashwin1111 — APPROVED** [review]
